File: languages/deno/multi.js
'use strict';

function lcgNext(state) {
  state.v = (Math.imul(1664525, state.v >>> 0) + 1013904223) >>> 0;
  return state.v >>> 0;
}

function sievePrimes(N) {
  if (N < 2) { console.log(0); return; }
  const comp = new Uint8Array(N + 1);
  comp[0] = 1; comp[1] = 1;
  const lim = Math.floor(Math.sqrt(N));
  for (let p = 2; p <= lim; p++) {
    if (comp[p] === 0) {
      for (let m = p * p; m <= N; m += p) comp[m] = 1;
    }
  }
  let cnt = 0;
  for (let i = 2; i <= N; i++) if (comp[i] === 0) cnt++;
  console.log(cnt);
}

function sortInts(N) {
  const arr = new Uint32Array(N);
  const s = { v: 123456789 };
  for (let i = 0; i < N; i++) arr[i] = lcgNext(s);
  // Convert to regular array for JS sort (stable, numeric)
  const a = Array.from(arr);
  a.sort((x, y) => x - y);
  const xorv = (a[0] ^ a[(N/2)|0] ^ a[N-1]) >>> 0;
  let sum = 0n;
  for (let i = 0; i < N; i++) sum += BigInt(a[i] >>> 0);
  console.log(`${xorv} ${sum & ((1n<<64n)-1n)}`);
}

function matmul(n) {
  const N = n|0;
  const A = new Float64Array(N*N);
  const B = new Float64Array(N*N);
  const C = new Float64Array(N*N);
  const s = { v: 123456789 };
  const inv = 1.0 / 4294967296.0;
  for (let i = 0; i < N*N; i++) A[i] = lcgNext(s) * inv;
  for (let i = 0; i < N*N; i++) B[i] = lcgNext(s) * inv;
  for (let i = 0; i < N; i++) {
    const row = i*N;
    for (let k = 0; k < N; k++) {
      const aik = A[row+k];
      const col = k*N;
      for (let j = 0; j < N; j++) {
        C[row+j] += aik * B[col+j];
      }
    }
  }
  let sum = 0.0;
  for (let i = 0; i < N*N; i++) sum += C[i];
  // print raw IEEE754 bits as hex
  const buf = new ArrayBuffer(8);
  const dv = new DataView(buf);
  dv.setFloat64(0, sum, false); // big-endian doesn't matter; we read as u64
  const hi = dv.getUint32(0, false), lo = dv.getUint32(4, false);
  const u = (BigInt(hi) << 32n) | BigInt(lo);
  console.log(u.toString(16).padStart(16,'0'));
}

function kmp(N, M) {
  const s = { v: 123456789 };
  const T = new Uint8Array(N);
  const P = new Uint8Array(M);
  for (let i = 0; i < N; i++) T[i] = 97 + (lcgNext(s) % 26);
  for (let i = 0; i < M; i++) P[i] = 97 + (lcgNext(s) % 26);
  const lps = new Int32Array(M);
  for (let i = 1, len = 0; i < M;) {
    if (P[i] === P[len]) { lps[i++] = ++len; }
    else if (len !== 0) { len = lps[len-1]; }
    else { lps[i++] = 0; }
  }
  let cnt = 0n, i = 0, j = 0;
  while (i < N) {
    if (T[i] === P[j]) { i++; j++; if (j === M) { cnt++; j = lps[j-1]; } }
    else if (j !== 0) { j = lps[j-1]; }
    else { i++; }
  }
  console.log(cnt.toString());
}

function main() {
  if (process.argv.length < 4) {
    console.error(`usage: node multi.js <task> <args...>`);
    process.exit(1);
  }
  const task = process.argv[2];
  if (task === 'sieve_primes') {
    const N = parseInt(process.argv[3],10);
    sievePrimes(N);
  } else if (task === 'sort_ints') {
    const N = parseInt(process.argv[3],10);
    sortInts(N);
  } else if (task === 'matmul_f64') {
    const n = parseInt(process.argv[3],10);
    matmul(n);
  } else if (task === 'string_kmp') {
    if (process.argv.length < 5) { console.error('need N M'); process.exit(1); }
    const N = parseInt(process.argv[3],10);
    const M = parseInt(process.argv[4],10);
    kmp(N,M);
  } else {
    console.error('unknown task');
    process.exit(1);
  }
}
main();

File: languages/pypy/multi.py
#!/usr/bin/env pypy3
import sys, math, array, struct

def lcg_next(s: int) -> int:
    return (1664525 * s + 1013904223) & 0xFFFFFFFF

def sieve_primes(N: int):
    if N < 2:
        print(0); return
    comp = bytearray(N+1)
    comp[0]=1; comp[1]=1
    lim = int(math.isqrt(N))
    for p in range(2, lim+1):
        if comp[p]==0:
            start = p*p
            comp[start:N+1:p] = b'\x01'*(((N-start)//p)+1)
    print((N-1) - sum(comp[2:]))

def sort_ints(N: int):
    s = 123456789
    a = [0]*N
    for i in range(N):
        s = lcg_next(s); a[i] = s
    a.sort()
    xorv = (a[0] ^ a[N//2] ^ a[-1]) & 0xFFFFFFFF
    total = 0
    for v in a:
        total = (total + (v & 0xFFFFFFFF)) & 0xFFFFFFFFFFFFFFFF
    print(f"{xorv} {total}")

def matmul(n: int):
    N=n
    s=123456789
    inv = 1.0/4294967296.0
    A=[0.0]*(N*N); B=[0.0]*(N*N); C=[0.0]*(N*N)
    for i in range(N*N):
        s = lcg_next(s); A[i] = (s)*inv
    for i in range(N*N):
        s = lcg_next(s); B[i] = (s)*inv
    for i in range(N):
        row=i*N
        for k in range(N):
            aik=A[row+k]
            col=k*N
            for j in range(N):
                C[row+j] += aik * B[col+j]
    sm = 0.0
    for v in C: sm += v
    bits = struct.unpack('>Q', struct.pack('>d', sm))[0]
    print(f"{bits:016x}")

def kmp(N: int, M: int):
    s = 123456789
    T = bytearray(N); P = bytearray(M)
    for i in range(N):
        s = lcg_next(s); T[i] = 97 + (s % 26)
    for i in range(M):
        s = lcg_next(s); P[i] = 97 + (s % 26)
    lps = [0]*M
    length = 0; i = 1
    while i < M:
        if P[i]==P[length]:
            length += 1; lps[i] = length; i += 1
        elif length != 0:
            length = lps[length-1]
        else:
            lps[i] = 0; i += 1
    cnt = 0
    i = j = 0
    while i < N:
        if T[i]==P[j]:
            i += 1; j += 1
            if j == M:
                cnt += 1; j = lps[j-1]
        elif j != 0:
            j = lps[j-1]
        else:
            i += 1
    print(cnt)

def main():
    if len(sys.argv) < 3:
        print(f"usage: {sys.argv[0]} <task> <args...>", file=sys.stderr)
        sys.exit(1)
    task = sys.argv[1]
    if task == "sieve_primes":
        sieve_primes(int(sys.argv[2]))
    elif task == "sort_ints":
        sort_ints(int(sys.argv[2]))
    elif task == "matmul_f64":
        matmul(int(sys.argv[2]))
    elif task == "string_kmp":
        if len(sys.argv) < 4:
            print("need N M", file=sys.stderr); sys.exit(1)
        kmp(int(sys.argv[2]), int(sys.argv[3]))
    else:
        print("unknown task", file=sys.stderr); sys.exit(1)

if __name__ == "__main__":
    main()